satpy/readers/osisaf_l3_nc.py

```diff
@@ -0,0 +1,241 @@
+                                   description="osisaf_polar_stereographic",
+                                   proj_id="osisaf_polar_stereographic",
+                                   projection=proj4str, width=x_size, height=y_size, area_extent=area_extent,
+                                   units="deg")
+        return area_def
+
+
+    def _get_finfo_grid(self):
+        """Get grid in case of filename info being used."""
+        if self.filename_info["grid"] == "ease":
+            self.area_def = self._get_ease_grid()
+            return self.area_def
+        elif self.filename_info["grid"] == "polstere" or self.filename_info["grid"] == "stere":
+            self.area_def = self._get_polar_stereographic_grid()
+            return self.area_def
+        else:
+            raise ValueError(f"Unknown grid type: {self.filename_info['grid']}")
+
+    def _get_ftype_grid(self):
+        """Get grid in case of filetype info being used."""
+        if self.filetype_info["file_type"] == "osi_radflux_grid":
+            self.area_def = self._get_geographic_grid()
+            return self.area_def
+        elif self.filetype_info["file_type"] == "osi_sst":
+            self.area_def = self._get_polar_stereographic_grid()
+            return self.area_def
+
+    def get_area_def(self, area_id):
+        """Override abstract baseclass method"""
+        if "grid" in self.filename_info:
+            return self._get_finfo_grid()
+        else:
+            return self._get_ftype_grid()
+
+
+    def _get_ds_attr(self, a_name):
+        """Get a dataset attribute and check it's valid."""
+        try:
+            return self[a_name]
+        except KeyError:
+            return None
+
+    def _get_ds_units(self, ds_info, var_path):
+        """Find the units of the datasets."""
+
+        file_units = ds_info.get("file_units")
+        if file_units is None:
+            file_units = self._get_ds_attr(var_path + "/attr/units")
+            if file_units is None:
+                file_units = 1
+        return file_units
+
+    def get_dataset(self, dataset_id, ds_info):
+        """Load a dataset."""
+        logger.debug(f"Reading {dataset_id['name']} from {self.filename}")
+        var_path = ds_info.get("file_key", f"{dataset_id['name']}")
+
+        shape = self[var_path + "/shape"]
+        data = self[var_path]
+        if shape[0] == 1:
+            # Remove the time dimension from dataset
+            data = data[0]
+
+        file_units = self._get_ds_units(ds_info, var_path)
+
+        # Try to get the valid limits for the data.
+        # Not all datasets have these, so fall back on assuming no limits.
+        valid_min = self._get_ds_attr(var_path + "/attr/valid_min")
+        valid_max = self._get_ds_attr(var_path + "/attr/valid_max")
+        if valid_min is not None and valid_max is not None:
+            data = data.where(data >= valid_min, np.nan)
+            data = data.where(data <= valid_max, np.nan)
```

Comment:
If the DataArray are floats then removing the `np.nan` should make xarray smart enough to pick the NaN with the right dtype. Otherwise you have to do `data.dtype.type(np.nan)`. Note I say this because of a bug currently in xarray when combined with numpy 2.x which is not released yet. This is a preventative suggestion.

Without this change if xarray doesn't fix their usage, then when numpy 2.0 comes out your dtype would suddenly get upcast to 64-bit floats. If the variable arrays are integers then you may also have to force the casting to 32-bit floats before this call.

```suggestion
            data = data.where(data >= valid_min)
            data = data.where(data <= valid_max)
```

Reply:
Code seems to work with your suggestion in place, so have found and removed all instances of `np.nan` in the `where` statements.